cpp/src/fil/internal.cuh

```diff
@@ -266,6 +295,122 @@ struct forest_params_t {
 /// FIL_TPB is the number of threads per block to use with FIL kernels
 const int FIL_TPB = 256;
 
+const uint32_t max_precise_int_float = 1 << 24;  // 16'777'216
+
+__host__ __device__ __forceinline__ bool fetch_bit(const uint8_t* array, int bit)
+{
+  const int bits_per_byte = 8;
+  return array[bit / bits_per_byte] & (1 << bit % bits_per_byte);
```

Comment:
`(array[bit / BITS_PER_BYTE] >> (bit % BITS_PER_BYTE)) & 1` is better, as it will always return 0 or 1 (and can have `int` as the return type).